### PR TITLE
Feature/inbound outbound peerlimit

### PR DESCRIPTION
**Pull request by munna0908**
# Description

This PR introduces two separate flags to customise the max-inbound-peers and max-outbound-peers limit for a node. 
Default values for max-inbound-peers & max-outbound-peers are `32` and `8` respectively. 
Users can also specify the limit either by using `max-peers` or `max-inbound/outbound-peers` flag
# Changes include

- [ ] Bugfix (non-breaking change that solves an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)

# Breaking changes

~~Removed the MaxPeers flag, so user need to specify separate limits for inbound and outbound peers~~
We have reintroduced `max-peers` to support backward compatibility. 

# Checklist

- [x] I have assigned this PR to myself
- [x] I have added at least 1 reviewer
- [x] I have tested this code
- [x] I have updated the README and other relevant documents (guides...)
- [ ] I have added sufficient documentation both in code, as well as in the READMEs

### CODE REVIEW

File: network/identity.go [hunk 1]
```diff
@@ -26,35 +26,52 @@ var (
...
 }
 
 func (i *identity) isPending(id peer.ID) bool {
-	val, ok := i.pending.Load(id)
+	_, ok := i.pending.Load(id)
 	if !ok {
 		return false
 	}
-	return val.(bool)
+	return ok
```
Comment (Kourin1996) [inline network/identity.go]: Why don't you return `ok` directly?

**Review by Kourin1996 — APPROVED** [review]
Looks good to me apart from small improvements

File: network/identity.go (hunk 1 above)
Comment (munna0908) [reply network/identity.go]: I missed it thanks Commit [link](https://github.com/0xPolygon/polygon-sdk/pull/326/commits/01bcc1cb9296c880f2846d10053b176db288f8e4)

**Comment by munna0908** [comment]
@Kourin1996 I have updated the Inbound Conns test case. Commit [link](https://github.com/0xPolygon/polygon-sdk/pull/326/commits/b66d2ed969afd2f0d6e09aae53c592046c3a1f67)

**Review by Kourin1996 — APPROVED** [review]
Looks good to me!

**Review by zivkovicmilos — CHANGES_REQUESTED** [review]
Please run a formatter and import organizer on all of the changed files, and re-request a review 👍

File: command/helper/config.go [hunk 2]
```diff
@@ -65,8 +66,9 @@ func DefaultConfig() *Config {
 		DataDir:        "./test-chain",
 		BlockGasTarget: "0x0", // Special value signaling the parent gas limit should be applied
 		Network: &Network{
-			NoDiscover: false,
-			MaxPeers:   50,
+			NoDiscover:       false,
+			MaxOutboundPeers: 10,
+			MaxInboundPeers:  40,
```
Comment (zivkovicmilos) [inline command/helper/config.go]: How did you derive that the default config should be 50 for outbound peers, and 10 for inbound?

File: network/discovery.go [hunk 3]
```diff
@@ -259,7 +259,7 @@ func (d *discovery) run() {
 
 func (d *discovery) handleDiscovery() {
 	// take a random peer and find peers
-	if target := d.peers.getRandomPeer(); target != nil {
+	if target := d.peers.getRandomPeer(); d.srv.numOpenSlots() > 0 && target != nil {
```
Comment (zivkovicmilos) [inline network/discovery.go]: Why would you get a random peer if you don't even check that you can dial them beforehand, but you check after you grabbed a random peer?

getRandomPeer is not a cheap operation as it involves locking and spinning up the random number generator

File: network/identity.go [hunk 4]
```diff
@@ -26,36 +26,48 @@ var (
...
 
 func (i *identity) delPending(id peer.ID) {
-	if _, loaded := i.pending.LoadAndDelete(id); loaded {
-		atomic.AddInt64(&i.pendingSize, -1)
+	if value, loaded := i.pending.LoadAndDelete(id); loaded {
+		if value.(network.Direction) == network.DirInbound {
+			atomic.AddInt64(&i.pendingInboundCount, -1)
+		} else if value.(network.Direction) == network.DirOutbound {
+			atomic.AddInt64(&i.pendingOutboundCount, -1)
+		}
```
Comment (zivkovicmilos) [inline network/identity.go]: Please break this if spaghetti into a switch so it's more readable

File: network/identity.go [hunk 5]
```diff
@@ -26,36 +26,48 @@ var (
...
-func (i *identity) setPending(id peer.ID) {
-	if _, loaded := i.pending.LoadOrStore(id, true); !loaded {
-		atomic.AddInt64(&i.pendingSize, 1)
+func (i *identity) setPending(id peer.ID, direction network.Direction) {
+	if _, loaded := i.pending.LoadOrStore(id, direction); !loaded {
+		if direction == network.DirInbound {
+			atomic.AddInt64(&i.pendingInboundCount, 1)
+		} else if direction == network.DirOutbound {
+			atomic.AddInt64(&i.pendingOutboundCount, 1)
+		}
```
Comment (zivkovicmilos) [inline network/identity.go]: Same here, please break it up into a switch statement for readability

File: network/identity.go [hunk 6]
```diff
@@ -87,13 +99,19 @@ func (i *identity) setup() {
...
+				i.srv.Disconnect(peerID, ErrNoAvailableSlots.Error())
+
+				return
+			}
+
+			if conn.Stat().Direction == network.DirInbound && i.srv.inboundConns() >= i.srv.maxInboundConns() {
 				i.srv.Disconnect(peerID, ErrNoAvailableSlots.Error())
 
 				return
 			}
```
Comment (zivkovicmilos) [inline network/identity.go]: Please make this a separate method that will do the disconnect logic if the slots are missing for the specific connection direction

File: network/server.go [hunk 7]
```diff
@@ -43,22 +44,24 @@ const (
...
 
 func DefaultConfig() *Config {
 	return &Config{
-		NoDiscover: false,
-		Addr:       &net.TCPAddr{IP: net.ParseIP("127.0.0.1"), Port: DefaultLibp2pPort},
-		MaxPeers:   10,
+		NoDiscover:       false,
+		Addr:             &net.TCPAddr{IP: net.ParseIP("127.0.0.1"), Port: DefaultLibp2pPort},
+		MaxOutboundPeers: 10,
+		MaxInboundPeers:  40,
```
Comment (zivkovicmilos) [inline network/server.go]: How did you come up with these values?

File: network/server.go [hunk 8]
```diff
@@ -377,14 +384,35 @@ func (s *Server) hasPeer(peerID peer.ID) bool {
...
 	}
 
 	return n
 }
 
+func (s *Server) inboundConns() int64 {
+	count := atomic.LoadInt64(&s.inboundConnCount)
+	if count < 0 {
+		count = 0
+	}
```
Comment (zivkovicmilos) [inline network/server.go:417]: Why would you need to check if count < 0? In what scenario would you have it < 0?

File: network/server.go [hunk 9]
```diff
@@ -419,7 +451,14 @@ func (s *Server) delPeer(id peer.ID) {
 	s.peersLock.Lock()
 	defer s.peersLock.Unlock()
 
-	delete(s.peers, id)
+	peer, ok := s.peers[id]
```
Comment (zivkovicmilos) [inline network/server.go]: You can inline this init within the if statement

File: network/server_test.go [hunk 10]
```diff
@@ -41,14 +41,14 @@ func TestConnLimit_Inbound(t *testing.T) {
...
 	}
 
-	// Server 2 tries to connect to Server 0
-	// but Server 0 is already connected to max peers
+	// Server 2 tries to connect to Server 1
+	// but Server 1 is already connected to max inbound peers
 	smallTimeout := time.Second * 5
-	if joinErr := JoinAndWait(servers[2], servers[0], smallTimeout, smallTimeout); joinErr == nil {
-		t.Fatal("Peer join should've failed")
+	if joinErr := JoinAndWait(servers[2], servers[1], smallTimeout, smallTimeout); joinErr == nil {
```
Comment (zivkovicmilos) [inline network/server_test.go:50]: Why did you change the servers being joined here? Server 0 should also be at max inbound peers after the previous join call in the test

File: network/server_test.go [hunk 11]
```diff
@@ -346,15 +347,18 @@ func TestNat(t *testing.T) {
 
 // TestPeerReconnection checks whether the node is able to reconnect with bootnodes on losing all active connections
 func TestPeerReconnection(t *testing.T) {
+	//Create bootnode config
```
Comment (zivkovicmilos) [inline network/server_test.go]: Nitpick warning: Please format the comment, add a space between // and the rest of the comment

**Review by zivkovicmilos — CHANGES_REQUESTED** [review]
I've left some comments that need to be resolved.

Overall it looks good 💯

File: command/helper/config.go (hunk 2 above)
Comment (munna0908) [reply command/helper/config.go]: Milos, it is 40 for inbound and 10 for outbound, 4:1 ratio is the preferred ratio and most of the protocols set 40:10 as default

**Comment by munna0908** [comment]
@zivkovicmilos , I have made all the suggested changes. Thanks

**Review by 0xpanoramix — APPROVED** [review]
LGTM 🚀

File: command/helper/config.go (hunk 2 above)
Comment (zivkovicmilos) [reply command/helper/config.go]: @Kourin1996 @lazartravica 
Can we discuss this?

File: network/discovery.go (hunk 3 above)
Comment (zivkovicmilos) [reply network/discovery.go]: If anybody reading this, this is the commit that fixes it:

https://github.com/0xPolygon/polygon-sdk/pull/326/commits/7fa13584e2cdc484c1189865c7b2a51bc7ecec5c

File: network/identity.go (hunk 4 above)
Comment (zivkovicmilos) [reply network/identity.go]: If anybody reading this, this is the commit that fixes it:

https://github.com/0xPolygon/polygon-sdk/pull/326/commits/7fa13584e2cdc484c1189865c7b2a51bc7ecec5c

File: network/identity.go (hunk 5 above)
Comment (zivkovicmilos) [reply network/identity.go]: If anybody reading this, this is the commit that fixes it:

https://github.com/0xPolygon/polygon-sdk/pull/326/commits/7fa13584e2cdc484c1189865c7b2a51bc7ecec5c

File: network/identity.go (hunk 6 above)
Comment (zivkovicmilos) [reply network/identity.go]: If anybody reading this, this is the commit that fixes it:

https://github.com/0xPolygon/polygon-sdk/pull/326/commits/7fa13584e2cdc484c1189865c7b2a51bc7ecec5c

File: network/server.go (hunk 9 above)
Comment (zivkovicmilos) [reply network/server.go]: If anybody reading this, this is the commit that fixes it:

https://github.com/0xPolygon/polygon-sdk/pull/326/commits/d5b6d98bc3210e3a5cc730859d015ceb6d63857a

File: network/server_test.go (hunk 11 above)
Comment (zivkovicmilos) [reply network/server_test.go]: If anybody reading this, this is the commit that fixes it:

https://github.com/0xPolygon/polygon-sdk/pull/326/commits/d5b6d98bc3210e3a5cc730859d015ceb6d63857a

File: network/server_test.go (hunk 10 above)
Comment (zivkovicmilos) [reply network/server_test.go:50]: Resolving this because Server 0 will have made an _outbound_ call to server 1, not _inbound_

**Review by zivkovicmilos — APPROVED** [review]
Thank you for resolving all of the conversations 🙏 

I've opened up a discussion regarding default limits, so we can all be on the same page 👍

File: command/helper/config.go [hunk 12]
```diff
@@ -63,8 +65,10 @@ func DefaultConfig() *Config {
 		DataDir:        "./test-chain",
 		BlockGasTarget: "0x0", // Special value signaling the parent gas limit should be applied
 		Network: &Network{
-			NoDiscover: false,
-			MaxPeers:   50,
+			NoDiscover:       false,
+			MaxPeers:         40,
```
Comment (zivkovicmilos) [inline command/helper/config.go:69]: Why the change from 50 to 40?

File: command/helper/helper.go [hunk 13]
```diff
@@ -476,13 +479,25 @@ func ReadConfig(baseCommand string, args []string) (*Config, error) {
 		return nil, err
 	}
 
+	if cliConfig.Network.MaxPeers != -1 {
```
Comment (zivkovicmilos) [inline command/helper/helper.go:482]: Where do you explicitly check that MaxInboundPeers and MaxOutboundPeers are set correctly?

Imagine I put that MaxInboundPeers is 5 and MaxOutboundPeers to be -1, where do you guard against that?

File: command/server/server_command.go [hunk 14]
```diff
@@ -143,14 +143,22 @@ func (c *ServerCommand) DefineFlags() {
 		FlagOptional: true,
 	}
 
-	c.FlagMap["max-peers"] = helper.FlagDescriptor{
-		Description: fmt.Sprintf("Sets the client's max peer count. Default: %d", helper.DefaultConfig().Network.MaxPeers),
+	c.FlagMap["max-inbound-peers"] = helper.FlagDescriptor{
```
Comment (zivkovicmilos) [inline command/server/server_command.go:155]: Where is the max-peers flag?

You've removed it, and only left max-outbound-peers and max-inbound-peers

File: command/helper/helper.go [hunk 15]
```diff
@@ -476,13 +479,25 @@ func ReadConfig(baseCommand string, args []string) (*Config, error) {
...
+		if cliConfig.Network.MaxInboundPeers != -1 || cliConfig.Network.MaxOutboundPeers != -1 {
+			return nil, errors.New(" both max-peers and max-inbound/outbound flags are set")
```
Comment (Kourin1996) [inline command/helper/helper.go]: you can remove first space in error message

File: command/helper/helper.go [hunk 16]
```diff
@@ -476,13 +479,25 @@ func ReadConfig(baseCommand string, args []string) (*Config, error) {
...
 	if configFile != "" {
 		// A config file has been passed in, parse it
 		diskConfigFile, err := readConfigFile(configFile)
 		if err != nil {
 			return nil, err
 		}
 
+		if diskConfigFile.Network.MaxPeers != -1 {
+			if diskConfigFile.Network.MaxInboundPeers != -1 || diskConfigFile.Network.MaxOutboundPeers != -1 {
+				return nil, errors.New(" both max-peers & max-inbound/outbound flags are set")
```
Comment (Kourin1996) [inline command/helper/helper.go]: you can remove first space in error message

**Review by zivkovicmilos — CHANGES_REQUESTED** [review]
Please resolve pending conversations, there are some major mistakes

**Comment by munna0908** [comment]
@zivkovicmilos We don't have an explicit check, because if user provides any limit <=-1, the default value is used. comment-[link](https://github.com/0xPolygon/polygon-edge/pull/326#discussion_r789025784)

**Comment by munna0908** [comment]
@zivkovicmilos Our limit was 40, When i am introducing the inbound and outbound flag, updated them to 40,10 respectively. But lazar suggested to retain the older limits, comment-[link](https://github.com/0xPolygon/polygon-edge/pull/326#discussion_r789018729)

**Comment by munna0908** [comment]
@zivkovicmilos Added the missing flag  commit-[link](https://github.com/0xPolygon/polygon-edge/pull/326/commits/375d5029ab160ece277498ccab80fad92948d200), comment-[link](https://github.com/0xPolygon/polygon-edge/pull/326#discussion_r789028253)

**Review by zivkovicmilos — APPROVED** [review]
Thank you for finally resolving all of the conversations 🙏 

Please check out @Kourin1996's comments.

Looks good to go now 👍

**Comment by munna0908** [comment]
@Kourin1996 removed whitespaces commit-[link](https://github.com/0xPolygon/polygon-edge/pull/326/commits/dc54add6d7c8d3a38d3e55725f55330b30d5b748)

**Review by lazartravica — APPROVED** [review]
